src/main/java/com/epam/esm/dao/impl/CertificateDaoImpl.java

```diff
@@ -0,0 +1,181 @@
+                return preparedStatement;
+            };
+            jdbcTemplate.update(preparedStatementCreator, keyHolder);
+
+            certificate.setId(keyHolder.getKey().longValue());
+        }
+        catch (DataAccessException e){
+            throw new DaoException(e);
+    }
+            return certificate;
+        }
+
+    @Transactional
+    @Override
+    public void delete(Certificate certificate) throws DaoException {
+        try {
+            jdbcTemplate.update(REMOVE_CERTIFICATE, certificate.getId());
+        } catch (DataAccessException e) {
+            throw new DaoException(e);
+        }
+    }
+
+    @Transactional
+    @Override
+    public void edit(Certificate certificate) throws DaoException {
+        certificate.setLastUpdateDate(new Timestamp(System.currentTimeMillis()));
+        final BeanPropertySqlParameterSource paramSource = new BeanPropertySqlParameterSource(certificate);
+        paramSource.registerSqlType("lastUpdateDate", Types.TIMESTAMP);
+        StringBuilder stringBuilder = new StringBuilder("UPDATE gift_certificate SET ")
+                .append("last_update_date =:lastUpdateDate");
+        Optional.ofNullable(certificate.getName()).ifPresent(x-> stringBuilder.append(", name =:name"));
+        Optional.ofNullable(certificate.getDescription()).ifPresent(x-> stringBuilder.append(", description =:description"));
+        Optional.ofNullable(certificate.getPrice()).ifPresent(x-> stringBuilder.append(", price =:price"));
+        Optional.ofNullable(certificate.getPrice()).ifPresent(x-> stringBuilder.append(", duration =:duration"));
+        stringBuilder.append(" where id =:id");
+        this.namedParameterJdbcTemplate.update(stringBuilder.toString(), paramSource);
+    }
+
+    @Override
+    public Certificate getByName(String name) {
+        try {
+            return jdbcTemplate.queryForObject(GET_CERTIFICATE_BY_NAME, new Object[]{name},
+                    new BeanPropertyRowMapper<>(Certificate.class));
+        } catch (EmptyResultDataAccessException e){
+            return null;
+        }
+    }
+
+    @Override
+    public Certificate getById(Long id) throws DaoException {
+        try {
+            return jdbcTemplate.queryForObject(GET_CERTIFICATE_BY_ID, new Object[]{id},
+                    new BeanPropertyRowMapper<>(Certificate.class));
+        } catch (EmptyResultDataAccessException e) {
+            return null;
+        }
+    }
+
+    @Override
+    public List<Certificate> getCertificateByTag(Tag tagName) throws DaoException {
+        List<Long> ids = getCertificateIdsByTagId(tagName.getId());
+        List<String> idsString = ids.stream().map((id) -> id.toString()).collect(Collectors.toList());
+        String idList = String.join(",", idsString);
+        return jdbcTemplate.query(GET_CERTIFICATES_BY_IDS, new Object[]{idList}, new CertificateDaoImpl());
+    }
+
+    @Override
+    public Certificate mapRow(ResultSet resultSet, int i) throws SQLException {
```

Comment:
I think it's better to merge this method with RowMapProvider.

Reply:
Where did you use variable "int i"?

Reply:
Merged method with RowMapProvider. Variable "int i" is a part of overriden method.